File: day7/day7.rb
module HandTypes
	## Hand type constants
	FIVEOFAKIND = 7
	FOUROFAKIND = 6
	FULLHOUSE = 5
	THREEOFAKIND = 4
	TWOPAIR = 3
	ONEPAIR = 2
	HIGHCARD = 1
end

class Hand
	include Comparable
	attr :cards
	## Actual values don't matter, just order. * is Joker, not used in this class, subclass converts J to *
	@@cardToValue = {"*" => 1, "2" => 2, "3" => 3, "4" => 4, "5" => 5, "6" => 6, "7" => 7, "8" => 8, "9" => 9, "T" => 10, "J" => 11, "Q" => 12, "K" => 13, "A" => 14}
	## Method to get hand type
	def get_hand_type
		cards_array = cards.chars
		cards_array.map! {|c| @@cardToValue[c]}.sort!
		case
		## only one type of card means 5oaK
		when cards_array.uniq.size == 1
			return HandTypes::FIVEOFAKIND
		## non-4oaK card could be first or last, so two types and 2nd == 4th means 4oaK
		when (cards_array.uniq.size == 2 and cards_array[1] == cards_array[3])
			return HandTypes::FOUROFAKIND
		## two types of card can only be 4oaK or FH, so if not 4oaK, must be FH
		when cards_array.uniq.size == 2 
			return HandTypes::FULLHOUSE
		## 3oaK could be arranged as AAABC, ABBBC or ABCCC, so check for each
		when (cards_array.uniq.size == 3 and (cards_array[0] == cards_array[2] or cards_array[1] == cards_array[3] or cards_array[2] == cards_array[4]))
			return HandTypes::THREEOFAKIND
		## three types of card can only be 3oaK or 2P, so if not 3oaK, must be 2P
		when cards_array.uniq.size == 3
			return HandTypes::TWOPAIR
		## four types of card can only be 1P
		when cards_array.uniq.size == 4
			return HandTypes::ONEPAIR
		## if none of above, must be HC
		else
			return HandTypes::HIGHCARD
		end
	end
	def <=>(other)
		thisType = get_hand_type
		otherType = other.get_hand_type
		cmpTypes = thisType <=> otherType
		if cmpTypes == 0
			cmpCards = 0
			i = 0
			while cmpCards == 0
				cmpCards = @@cardToValue[@cards[i]] <=> @@cardToValue[other.cards[i]]
				i += 1
			end
			return cmpCards
		end
		return cmpTypes
	end
	def initialize(str)
		@cards = str
	end
	def inspect
		@cards
	end
end

class JokerHand < Hand
	def get_hand_type
		cards_array = cards.chars
		cards_array.map! {|c| @@cardToValue[c]}.sort!
		jokers = cards_array.include?(1)
		## Convert Joker to other card that gives best hand
		## If all cards are Jokers, always 5oaK
		if cards_array[4] == 1
			return HandTypes::FIVEOFAKIND
		## If 4 jokers, all will convert to last card, always 5oaK
		elsif cards_array[3] == 1
			return HandTypes::FIVEOFAKIND
		## If 3 jokers, convert all to highest card, then check normally
		elsif cards_array[2] == 1
			cards_array[0] = cards_array[4]
			cards_array[1] = cards_array[4]
			cards_array[2] = cards_array[4]
		## 2 Jokers
		elsif cards_array[1] == 1 
			## If 2 jokers, and two of lowest non-joker card, convert jokers to said card, in case of 4oaK
			if cards_array[2] == cards_array[3]
				cards_array[0] = cards_array[2]
				cards_array[1] = cards_array[2]
			## Otherwise, convert to highest
			else
				cards_array[0] = cards_array[4]
				cards_array[1] = cards_array[4]
			end
		# 1 Joker
		elsif cards_array[0] == 1
			## If 1 joker, and two of lowest non-joker card, but not two of highest card, i.e. JAABC or JAAAB, convert to lowest non-joker card
			if cards_array[1] == cards_array[2] and cards_array[3] != cards_array[4]
				cards_array[0] = cards_array[1]
			## If card is of form JABBC, convert to B for 3oaK (technically also catches JAAAA, but J can be converted to any card in this case)
			elsif cards_array[2] == cards_array[3]
				cards_array[0] = cards_array[2]
			## Otherwise convert to highest
			else
				cards_array[0] = cards_array[4]
			end
		end
		## Sort again after converting Jokers
		cards_array.sort!
		case
		## only one type of card means 5oaK
		when cards_array.uniq.size == 1
			return HandTypes::FIVEOFAKIND
		## non-4oaK card could be first or last, so two types and 2nd == 4th means 4oaK
		when (cards_array.uniq.size == 2 and cards_array[1] == cards_array[3])
			return HandTypes::FOUROFAKIND
		## two types of card can only be 4oaK or FH, so if not 4oaK, must be FH
		when cards_array.uniq.size == 2 
			return HandTypes::FULLHOUSE
		## 3oaK could be arranged as AAABC, ABBBC or ABCCC, so check for each
		when (cards_array.uniq.size == 3 and (cards_array[0] == cards_array[2] or cards_array[1] == cards_array[3] or cards_array[2] == cards_array[4]))
			return HandTypes::THREEOFAKIND
		## three types of card can only be 3oaK or 2P, so if not 3oaK, must be 2P
		when cards_array.uniq.size == 3
			return HandTypes::TWOPAIR
		## four types of card can only be 1P
		when cards_array.uniq.size == 4
			return HandTypes::ONEPAIR
		## if none of above, must be HC
		else
			return HandTypes::HIGHCARD
		end
	end
	def initialize(str)
		@cards = str.tr("J", "*")
	end
end

def get_winnings(jokers = false)
	totalWinnings = 0
	File.open("day7Input.txt") do |f|
		rounds = Array.new
		f.each_line(chomp: true) do |line|
			rounds.push(line.split(" "))
		end
		if jokers
			rounds.map! {|r| [JokerHand.new(r[0]), Integer(r[1])]}
		else
			rounds.map! {|r| [Hand.new(r[0]), Integer(r[1])]}
		end
		rounds.sort_by! {|r| r[0]}
		multiplier = 1
		rounds.each do |r|
			totalWinnings += r[1] * multiplier
			multiplier += 1
		end
	end
	return totalWinnings
end

p get_winnings
p get_winnings(true)

File: day07/day7.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strings"
	"strconv"
	"slices"
	"cmp"
)

const (
	HighCard = iota
	OnePair
	TwoPair
	ThreeOfAKind
	FullHouse
	FourOfAKind
	FiveOfAKind
)
	
var cardValuesMap = map[rune]int{'*': 1, '2': 2, '3': 3, '4': 4, '5': 5, '6': 6, '7': 7, '8': 8, '9': 9, 'T': 10, 'J': 11, 'Q': 12, 'K': 13, 'A': 14}

func get_hand_type(cards string) int {
	cardsValues := make([]int, len(cards))
	for i, r := range []rune(cards) {
		cardsValues[i] = cardValuesMap[r]
	}
	slices.Sort(cardsValues)
	// Handle Jokers
	// If 4 or more Jokers, always 5oaK
	switch {
	case cardsValues[3] == 1:
		return FiveOfAKind
	// If 3 Jokers, convert all to highest card and continue
	case cardsValues[2] == 1:
		cardsValues[0] = cardsValues[4]
		cardsValues[1] = cardsValues[4]
		cardsValues[2] = cardsValues[4]
	// If 2 Jokers, check for potential 4oaK, otherwise convert to highest
	case cardsValues[1] == 1:
		if cardsValues[2] == cardsValues[3] {
			cardsValues[0] = cardsValues[2]
			cardsValues[1] = cardsValues[2]
		} else {
			cardsValues[0] = cardsValues[4]
			cardsValues[1] = cardsValues[4]
		}
	// 1 Joker
	case cardsValues[0] == 1:
		// If 1 Joker, and at least two of lowest non-Joker card, but not two of highest card, i.e. hand is JAABC or JAAAB, convert Joker to lowest non-Joker card
		if cardsValues[1] == cardsValues[2] && cardsValues[3] != cardsValues[4] {
			cardsValues[0] = cardsValues[1]
		// If hand is JABBC, convert J to B for 3oaK
		} else if cardsValues[2] == cardsValues[3] {
			cardsValues[0] = cardsValues[2]
		// Otherwise convert to highest
		} else {
			cardsValues[0] = cardsValues[4]
		}
	}
	// Sort again after joker conversion
	slices.Sort(cardsValues)
	// Get type of card
	numCardTypes := len(slices.Compact(slices.Clone(cardsValues)))
	switch {
	// Only one type of card means 5oaK
	case numCardTypes == 1:
		return FiveOfAKind
	// If two types of cards and 2nd and 4th are the same, must be 4oaK, since cards are sorted
	case numCardTypes == 2 && cardsValues[1] == cardsValues[3]:
		return FourOfAKind
	// If only two types of cards, but not 4oaK, must be FH
	case numCardTypes == 2:
		return FullHouse
	// If three types of cards, check for 3oaK
	case numCardTypes == 3 && (cardsValues[0] == cardsValues[2] || cardsValues[1] == cardsValues[3] || cardsValues[2] == cardsValues[4]):
		return ThreeOfAKind
	// If not 3oaK, must be 2Pair
	case numCardTypes == 3:
		return TwoPair
	// If 4 types of cards, must be 1Pair
	case numCardTypes == 4:
		return OnePair
	// If none of the above, must be HC
	default:
		return HighCard
	}
}

type round struct {
	hand hand
	bid int
}

type hand struct {
	cards string
	handType int
}

func print_error(err error) {
	if err != nil {
		fmt.Println(err)
	}
}

func get_winnings(jokers bool) {
	file, err := os.Open("day7Input.txt")
	print_error(err)
	defer file.Close()

	rounds := []round{}
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		lineParts := strings.Split(line, " ")
		bid, err := strconv.Atoi(lineParts[1])
		print_error(err)
		cardsStr := lineParts[0]
		if jokers {
			cardsStr = strings.ReplaceAll(cardsStr, "J", "*")
		}
		hand := hand{cards: cardsStr, handType: get_hand_type(cardsStr)}
		newRound := round{hand: hand, bid: bid}
		rounds = append(rounds, newRound)
	}
	cmpHands := func (x, y round) int {
		cmpTypes := cmp.Compare(x.hand.handType, y.hand.handType)
		if cmpTypes == 0 {
			cmpCards := 0
			i := 0
			for (cmpCards == 0) && (i < len(x.hand.cards)) {
				cmpCards = cmp.Compare(cardValuesMap[rune(x.hand.cards[i])], cardValuesMap[rune(y.hand.cards[i])])
				i += 1
			}
			return cmpCards
		} else {
			return cmpTypes
		}
	}
	slices.SortFunc(rounds, cmpHands)
	totalWinnings := 0
	multiplier := 1
	for _, round := range rounds {
		totalWinnings += round.bid * multiplier
		multiplier += 1
	}
	fmt.Println(totalWinnings)
}

func main() {
	get_winnings(false)
	get_winnings(true)
}